_posts/2022-07-07-premature-dispsosal.md

```diff
@@ -0,0 +1,87 @@
+
+## Unpredictable behavior
+
+Whether `DoSomething` runs into this problem depends on what it does internally. As with throwing Exceptions, the behavior between methods that return Tasks but are actually synchronous (e.g. `Task SyncMethod() => Task.CompletedTask;`) and truly asynchronous methods (e.g. `async Task AsyncMethod() => await Task.Yield();`) varies. Consider these two implementations of `DoSomething`:
+
+```csharp
+static Task DoSomething(MyDisposableClass myDisposableClass)
+{
+    myDisposableClass.DoStuff();
+    return Task.CompletedTask;
+}
+```
+
+and
+
+```csharp
+static async Task DoSomething(MyDisposableClass myDisposableClass)
+{
+    await Task.Yield();
+
+    myDisposableClass.DoStuff();
+}
+```
+
+The first implementation will run without any exception in the intially shown `Main` method because `DoSomething` will run to completion before it returns to `Main` to dispose the passed object reference. The second version will throw an `ObjectDisposedException` because `Task.Yield()` will trigger the return to the caller and finish the remaining code of the asynchronous method later. We need to assume that any real `Task` returning method will have actual asynchronous code exception (like the 2nd version) though. In reality, a single method can even show both behaviors, completely depending on runtime conditions:
+
+```csharp
+static Task DoSomething(MyDisposableClass myDisposableClass)
+{
+    if (myDisposableClass.Validate)
+    {
+        return InvokeWithExpensiveValidation();
+
+        async Task InvokeWithExpensiveValidation()
+        {
+            await ExpensiveValidationAsync();
+            myDisposableClass.DoStuff();
+        }
+    }
+    else
+    {
+        myDisposableClass.DoStuff();
+        return Task.CompletedTask;
+    }
+}
+```
+
+An implementation like this can avoid the cost of asynchronous methods when it is expected that the async paths is only rarely used. This makes it even harder to notice the initial mistake to not `await` the method call because the code might work just fine most of time time, until it doesn't.
```

Comment:
```suggestion
An implementation like the presented one can avoid the cost of asynchronous methods when it is expected that the async path is only rarely used. However, it makes it even harder to notice the initial mistake of not `await`ing the method call because the code might work just fine most of times. Until it doesn't.
```